include/oneapi/dpl/pstl/utils.h

```diff
@@ -669,6 +559,151 @@ __dpl_ceiling_div(_T1 __number, _T2 __divisor)
+            __n -= __cur + 1, __first = ++__it;
+        }
+        else
+            __n = __cur;
+    }
+    return __first;
+}
+
+template <typename _Acc, typename _Size1, typename _Value, typename _Compare>
+_Size1
+__pstl_upper_bound(_Acc __acc, _Size1 __first, _Size1 __last, const _Value& __value, _Compare __comp)
+{
+    return __pstl_lower_bound(__acc, __first, __last, __value,
+                              oneapi::dpl::__internal::__not_pred<oneapi::dpl::__internal::__reorder_pred<_Compare>>{
+                                  oneapi::dpl::__internal::__reorder_pred<_Compare>{__comp}});
+}
+
+// Searching for the first element strongly greater than a passed value - right bound
+template <typename _Buffer, typename _Index, typename _Value, typename _Compare>
+_Index
+__pstl_right_bound(_Buffer& __a, _Index __first, _Index __last, const _Value& __val, _Compare __comp)
+{
+    return __pstl_upper_bound(__a, __first, __last, __val, __comp);
+}
+
+template <typename _IntType, typename _Acc>
+struct _ReverseCounter
+{
+    typedef ::std::make_signed_t<_IntType> difference_type;
+
+    _IntType __my_cn;
+
+    _ReverseCounter&
+    operator++()
+    {
+        --__my_cn;
+        return *this;
+    }
+
+    template <typename _DiffType>
+    _ReverseCounter&
+    operator+=(_DiffType __val)
+    {
+        __my_cn -= __val;
+        return *this;
+    }
+
+    difference_type
+    operator-(const _ReverseCounter& __a)
+    {
+        return __a.__my_cn - __my_cn;
+    }
+
+    operator _IntType() { return __my_cn; }
+
+// TODO: Temporary hotfix. Investigate the necessity of _ReverseCounter
+// Investigate potential user types convertible to integral
+// This is the compile-time trick where we define the conversion operator to sycl::id
+// conditionally. If we can call accessor::operator[] with the type that converts to the
+// same integral type as _ReverseCounter (it means that we can call accessor::operator[]
+// with the _ReverseCounter itself) then we don't need conversion operator to sycl::id.
+// Otherwise, we define conversion operator to sycl::id.
+#if _ONEDPL_BACKEND_SYCL
+    struct __integral
+    {
+        operator _IntType();
+    };
+
+    template <typename _Tp>
+    static auto
+    __check_braces(int) -> decltype(::std::declval<_Tp>()[::std::declval<__integral>()], ::std::false_type{});
+
+    template <typename _Tp>
+    static auto
+    __check_braces(...) -> ::std::true_type;
+
+    class __private_class;
+
+    operator ::std::conditional_t<decltype(__check_braces<_Acc>(0))::value, sycl::id<1>, __private_class>()
+    {
+        return sycl::id<1>(__my_cn);
+    }
+#endif
+};
+
+// Reverse searching for the first element strongly less than a passed value - left bound
+template <typename _Buffer, typename _Index, typename _Value, typename _Compare>
+_Index
+__pstl_left_bound(_Buffer& __a, _Index __first, _Index __last, const _Value& __val, _Compare __comp)
+{
+    auto __beg = _ReverseCounter<_Index, _Buffer>{__last - 1};
```

Comment:
Revere counter iterator looks like a little bit strange for me,
As far as I know we already have `oneapi::dpl::counting_iterator`. Is it not enough?
Also when I see the API like `_Index __first, _Index __last` I assume that the element before `__first` isn't exist and the element iterated by `__last` isn't exist too. From this point of view what does it mean `__first - 1` below?

@danhoeflinger what do you think about this?

Reply:
@SergeyKopienko and I discussed offline and agreed. This code is just being moved so that we can group these similar utilities together.

If there is an issue or room for improvement in `__pstl_left_bound`, then we can open a GH issue to address in a future PR.

Reply:
So, @mmichel11, could you please create the issue for this point?